Telemetry Refactoring and Stable ClientId

vxfield opened the pull request:
This PR aims to fix two telemetry issues:

**1) Telemetry was missing some events**
The telemetry was not capturing some events because the telemetry class was being initialized and subscribed to Dependency Injection services, like References package loaded, after the events were being triggered. This was necessary to capture the MetadataChanged events that is used on issue number 2.

The proposed fix:
1) Refactor the telemetry class into a dependency injection service, like all other services.
2) Create and add the `ITelemetryService` in all services' constructor to make sure it gets initialized before any other service. There are other ways to do this, but it would need a change in JupyterCore to force a new instance of ITelemetryService to be created during the App initialization, like calling `public void Configure(IApplicationBuilder app, IWebHostEnvironment env)` from `Startup.cs`.
3) Since the services depend on the `ITelemetryService` (for it to get instantiate) and the `TelemetryService` depends on the services (to subscribe to their events), it would cause a circular dependency which is not allowed.
So I created a PubSub Event Service to avoid the circular dependency. This service is very generic and could also be used for communication between services in an event-based fashion.

**2) Having an unstable DeviceId impacts the count of unique users in transient environments**
In environments like MyBinder/Kubernetes/Docker where a container is instantiated and disposed several times on different machines, the DeviceId is unstable and will lead to duplicated users count.

The proposed fix:
1) When telemetry is enabled (not removed during compilation, or not opt-ed out) we persist an anonymous (randomly generated) ClientId in the browser's cookie storage.
2) The `Kernel.js` first check if the telemetry is not opted-out, and if not, it will call a service that:
      1. Checks if we need the user's consent to store non-essential data in the cookie (complying with GDPR regulations);
      2. If consent is not needed or if it's granted, the service generates a random guid and stores it in the cookie storage.

vxfield commented:
I'm holding on updating the unit tests to see if you agree with my refactoring of the telemetry as a DI service.
Once you are okay with that approach, I'll update the unit tests (by instantiating the DummyTelemetryService).

cgranade commented:
Thanks for tagging me! Reviewing now, will take a little bit, but in the meantime I wanted to respond to this real quick:

> 2. Create and add the `ITelemetryService` in all services' constructor to make sure it gets initialized before any other service. There are other ways to do this, but it would need a change in JupyterCore to force a new instance of ITelemetryService to be created during the App initialization, like calling `public void Configure(IApplicationBuilder app, IWebHostEnvironment env)` from `Startup.cs`.

It's not a bad idea to give additional hooks to microsoft/jupyter-core. If it would be helpful to have a configuration action there as well, let's go on and add that.

vxfield commented:
Thanks. That would be great.
You do it or I do it? If it’s the former, you can just create the infra and then I adapt my code.

________________________________
From: Chris Granade <notifications@github.com>
Sent: Sunday, March 8, 2020 10:50:58 AM
To: microsoft/iqsharp <iqsharp@noreply.github.com>
Cc: XField <vxfield@outlook.com>; Author <author@noreply.github.com>
Subject: Re: [microsoft/iqsharp] Telemetry Refactoring and Stable ClientId (#101)


Thanks for tagging me! Reviewing now, will take a little bit, but in the meantime I wanted to respond to this real quick:

  1.  Create and add the ITelemetryService in all services' constructor to make sure it gets initialized before any other service. There are other ways to do this, but it would need a change in JupyterCore to force a new instance of ITelemetryService to be created during the App initialization, like calling public void Configure(IApplicationBuilder app, IWebHostEnvironment env) from Startup.cs.

It's not a bad idea to give additional hooks to microsoft/jupyter-core. If it would be helpful to have a configuration action there as well, let's go on and add that.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fiqsharp%2Fpull%2F101%3Femail_source%3Dnotifications%26email_token%3DAN3JLUK5BKIWKCQBYXDT7D3RGPSJFA5CNFSM4LDXYSC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOE4O4Q%23issuecomment-596232050&data=02%7C01%7C%7Cc0f4bd42afce466f557e08d7c38943a6%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637192866616772065&sdata=JEofjyIWr9i2G0sOb6EtQEBlaIGrFCfiHIhJNKVZQsY%3D&reserved=0>, or unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAN3JLUNCIUM5XPGMTPNUIKDRGPSJFANCNFSM4LDXYSCQ&data=02%7C01%7C%7Cc0f4bd42afce466f557e08d7c38943a6%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637192866616782070&sdata=TzGta2jZnXbDqjng4O5fLoq%2FBw9iRNZEhOA4akM8F8Y%3D&reserved=0>.

cgranade commented on src/Core/Events/EventService.cs:
```diff
@@ -0,0 +1,173 @@
+        /// </summary>
+        /// <param name="eventType">The type of the event</param>
+        /// <param name="action">The action (of type Action<TArg>) to be executed when the event is triggered</param>
+        public void Subscribe(Type eventType, object action)
```
Alternatively, could make `eventType` a type parameter using the same `where` trick as at L74, then use `typeof` when adding back into the dictionary.

cgranade commented on src/Core/Events/EventService.cs:
```diff
@@ -0,0 +1,173 @@
+    /// A basic multi-thread implementation of a publisher-subscriber based on
+    /// the type of the event.
+    /// </summary>
+    public class EventPubSub
```
It feels like this is an internal-facing class, since the typed wrapper is what's used directly?

vxfield commented on src/Jupyter/client/telemetry.ts:
```diff
@@ -0,0 +1,194 @@
+
+interface ConsentMarkup {
+    Markup: string;
+    Javascripts: string[];
```
When we call the service API the service returns a ClientInfo object. As ClientInfo gets deserialized (line 153) so does thee ConsentMarkup

vxfield commented on src/Jupyter/client/telemetry.ts:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    public requestConsent(consentMarkup: ConsentMarkup) {
```
At this point the service has already been called, so the user had connection.
The CookieConsentHelper don't call any services, it's just do work on the client to show or not show the banner, and set a cookie if consent was granted.
But agree that would be good to have error handling, likely on getClientInfoAsync.

vxfield commented on src/Jupyter/client/telemetry.ts:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    private onCssLoaded(consentHelper: CookieConsentHelper) {
```
Great idea. I'll do that.

vxfield commented on src/Jupyter/client/telemetry.ts:
```diff
@@ -0,0 +1,194 @@
+        if (consentHelper.countJsLoaded == consentHelper.consentMarkup.Javascripts.length) {
+            console.log("Calling Cookie API hasConsent");
+            this.mscc = mscc;
+            this.mscc.on('consent', consentHelper.onConsentGranted);
```
The MSCC (Microsoft Cookie Compliance) javascript injects code in the DOM to detect user's consent when the user engages with the site (for instance, by clicking on a link).
When the events considered by MSCC to be engagement happens, then it triggers this "consent" event.

vxfield commented on src/Tool/Telemetry.cs:
```diff
@@ -122,6 +93,23 @@ public static void InitLogManager(IConfiguration config)
                 .Where(address => address != null && !string.IsNullOrWhiteSpace(address) && !address.StartsWith("000000"))
                 .FirstOrDefault();
+
+        public void SetSharedContextIfChanged(IMetadataController metadataController, string propertyChanged, params string[] propertyWhitelist)
```
Agree

vxfield commented:
Thanks! Now I only need to do the PR for the service part, deploy it to prod, and update the telemetry.ts with the production URL before we merge.

cgranade commented:
> Thanks! Now I only need to do the PR for the service part, deploy it to prod, and update the telemetry.ts with the production URL before we merge.

May I suggest merging on to a feature branch, then, so that you can easily get beta builds from this to use with the other parts of this feature?

cgranade commented on src/Tests/Startup.cs:
```diff
@@ -43,6 +43,7 @@ internal static ServiceProvider CreateServiceProvider(string workspaceFolder)
         public static void AddMocks(this IServiceCollection services)
         {
             var shell = new MockShell();
+            services.AddSingleton<ITelemetryService,NullTelemetryService>();
```
```suggestion
            services.AddSingleton<ITelemetryService, NullTelemetryService>();
```

vxfield commented:
@cgranade, I've addressed all the previous feedback, and updated the client id API URL to production.
After you review my latest changes to instrument the origin, I think we can merge it to master.
